File: Support/Templates/cs/Files/EditableRootParent.cs
using System;
using Csla;

namespace Templates
{
  [Serializable]
  public class EditableRootParent : BusinessBase<EditableRootParent>
  {
    public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(nameof(Id));
    public int Id
    {
      get { return GetProperty(IdProperty); }
      set { SetProperty(IdProperty, value); }
    }

    public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
    public string Name
    {
      get { return GetProperty(NameProperty); }
      set { SetProperty(NameProperty, value); }
    }

    // example with Editable Child list 
    public static readonly PropertyInfo<EditableChildList> ChildListProperty = RegisterProperty<EditableChildList>(nameof(ChildList), "Child list");
    public EditableChildList ChildList
    {
      get { return GetProperty<EditableChildList>(ChildListProperty); }
    }

    // Example with Editable Child 
    public static readonly PropertyInfo<EditableChild> ChildProperty = RegisterProperty<EditableChild>(nameof(Child), "Child");
    public EditableChild Child
    {
      get { return GetProperty<EditableChild>(ChildProperty); }
    }

    protected override void AddBusinessRules()
    {
      // TODO: add validation rules
      //BusinessRules.AddRule(new Rule(), IdProperty);
    }

    private static void AddObjectAuthorizationRules()
    {
      // TODO: add authorization rules
      //BusinessRules.AddRule(...);
    }

    [RunLocal]
    [Create]
    private void Create()
    {
      // TODO: load default values
      // omit this override if you have no defaults to set
      LoadProperty(ChildListProperty, DataPortal.CreateChild<EditableChildList>());
      LoadProperty(ChildProperty, DataPortal.CreateChild<EditableChild>());
      BusinessRules.CheckRules();
    }

    [Fetch]
    private void Fetch(int id)
    {
      // TODO: load values
      LoadProperty(ChildListProperty, DataPortal.FetchChild<EditableChildList>(null));
      LoadProperty(ChildProperty, DataPortal.FetchChild<EditableChild>(null));
    }

    [Insert]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Insert()
    {
      // TODO: insert values
      FieldManager.UpdateChildren(this);
    }

    [Update]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Update()
    {
      // TODO: update values
      FieldManager.UpdateChildren(this);
    }

    [DeleteSelf]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void DeleteSelf()
    {
      Delete(this.Id);
    }

    [Delete]
    [Transactional(TransactionalTypes.TransactionScope)]
    private void Delete(int id)
    {
      // TODO: delete values
      FieldManager.UpdateChildren(this);
    }
  }
}

File: Support/Templates/vb/Files/EditableRootParent.vb
Imports System
Imports Csla
Imports Csla.Security

<Serializable()>
Public Class EditableRootParent
  Inherits BusinessBase(Of EditableRootParent)

  Public Shared ReadOnly IdProperty As PropertyInfo(Of Integer) = RegisterProperty(Of Integer)(NameOf(Id))
  Public Property Id() As Integer
    Get
      Return GetProperty(IdProperty)
    End Get
    Set(ByVal value As Integer)
      SetProperty(IdProperty, value)
    End Set
  End Property

  Public Shared ReadOnly NameProperty As PropertyInfo(Of String) = RegisterProperty(Of String)(NameOf(Name))
  Public Property Name() As String
    Get
      Return GetProperty(NameProperty)
    End Get
    Set(ByVal value As String)
      SetProperty(NameProperty, value)
    End Set
  End Property

  Public Shared ReadOnly ChildListProperty As PropertyInfo(Of EditableChildList) = RegisterProperty(Of EditableChildList)(NameOf(ChildList), "Child list")
  Public ReadOnly Property ChildList() As EditableChildList
    Get
      Return GetProperty(Of EditableChildList)(ChildListProperty)
    End Get
  End Property

  Public Shared ReadOnly ChildProperty As PropertyInfo(Of EditableChild) = RegisterProperty(Of EditableChild)(NameOf(Child), "Child")
  Public ReadOnly Property Child() As EditableChild
    Get
      Return GetProperty(Of EditableChild)(ChildProperty)
    End Get
  End Property

  Protected Overrides Sub AddBusinessRules()
    'call base class implementation to add data annotation rules to BusinessRules 
    MyBase.AddBusinessRules()

    ' TODO: add validation rules
    'BusinessRules.AddRule(New MyRule, IdProperty)

  End Sub

  Public Shared Sub AddObjectAuthorizationRules()
    'TODO: add authorization rules
    'BusinessRules.AddRule(...)
  End Sub

  <Create>
  <RunLocal()>
  Private Sub Create()
    ' TODO: load default values 
    ' omit this override if you have no defaults to set 
    LoadProperty(ChildListProperty, DataPortal.CreateChild(Of EditableChildList)())
    LoadProperty(ChildProperty, DataPortal.CreateChild(Of EditableChild)())
    BusinessRules.CheckRules()
  End Sub

  <Fetch>
  Private Sub Fetch(ByVal id As Integer)
    ' TODO: load values 
    LoadProperty(ChildListProperty, DataPortal.FetchChild(Of EditableChildList)(Nothing))
    LoadProperty(ChildProperty, DataPortal.FetchChild(Of EditableChild)(Nothing))
  End Sub

  <Insert>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Sub Insert()
    ' TODO: insert values 
    FieldManager.UpdateChildren(Me)
  End Sub

  <Update>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Sub Update()
    ' TODO: update values 
    FieldManager.UpdateChildren(Me)
  End Sub

  <DeleteSelf>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Sub DeleteSelf()
    Delete(Me.Id)
  End Sub

  <Delete>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Shadows Sub Delete(ByVal id As Integer)
    ' TODO: delete values 
    FieldManager.UpdateChildren(Me)
  End Sub

End Class
